README.md
# ReadAxfBOM

[![Build Status](https://travis-ci.org/MarkusLoew/MyFits.svg?branch=master)](https://travis-ci.org/MarkusLoew/MyFits)

R package to add either a inverse-quadratic, bell-shaped, or "waiting-in-line" curve fit to data.

See 

	help(package = "MyFits") 

for details on the function provided by this package.
The import functions returns either a list of curve fit coefficients and r2, or the nls-model, or the predicted values.


### Installation

Installation straight from github (if package "devtools" is already installed) via

```{r}
devtools::install_github("MarkusLoew/MyFits")
```

### Example session
```{r}
d <- data.frame(x = 1:10,
                 y = c(0.5, 2, 10, 9, 8, 6.5, 5, 4, 3, 2.5))
MyFits(x = x, y = y, data = d)
MyFits(x = x, y = y, data = d, fit = "Wait")
nls.fit <- MyFits(x = x, y = y, data = d, fit = "Bell", model = TRUE)
my.x <- seq(min(d$x), max(d$x), length.out = 100)
my.predicted <- predict(nls.fit, data.frame(x = my.x))
mod.pred <- data.frame(x = my.x,
                       y = my.predicted)
plot(d)
lines(y ~ x, data = mod.pred)

require(ggplot2)
p <- ggplot(d, aes(x = x, y = y))
 p <- p + geom_point()
 p <- p + stat_myfits(fit = "Inverse_quad", colour = "blue")
 p <- p + stat_myfits(fit = "Bell", colour = "red")
 p <- p + stat_myfits(fit = "Wait", colour = "green")
p 
```


NEWS
MyFits 0.0.0.3
----------------------------------------------------------------

* Updated data argument handling in nls

MyFits 0.0.0.2
----------------------------------------------------------------

* Allows providing custom starting values to the fit


MyFits 0.0.0.1
----------------------------------------------------------------

* Initial version



R/MyFits.R
#' Fits either an inverse quadratic, a bell-shaped or a waiting-in-line curve to data.
#' @description Fits one of three curves to x, y data using \code{nls}. Available fitting functions are inverse quadratic, bell-shaped, or waiting-in-line.
#' @param x x values
#' @param y y values 
#' @param data data frame with the x and y values
#' @param fit Name of the fit equation to be used. Either \code{Inverse_quad} for an inverse quadratic equation \code{y ~ x / (a + b*x + c*x^2)}, \code{Bell} for a bell shaped curve \code{y ~ amp * exp(-0.5*((x - xmean)/xsd)^2)}, or \code{Wait} for waiting-in-line \code{y ~ A * kw * x * exp(-kw*x)}, after Ritchie 2008).
#' @param model Logical. Should the \code{nls} model be returned? Default is FALSE, only coefficient and R2 are returned.
#' @param coords Logical. If \code{TRUE} the model predictions are returned as data frame. Usful for plotting the model result. See \code{stat_MyFits} for a ggplot2-implementation.
#' @param start.a optional initial guess for the curve fit coefficients \code{a}, \code{amp}, or \code{A} (depends on which equation was selected). Default is NA, i.e. an internal guess is used to determine \code{start.a}.
#' @param start.b optional initial guess for the second curve fit coefficient for the \code{b}, \code{xmean} or \code{kw} parameter. Default is NA, i.e. an internal guess is used to determine \code{start.b}.
#' @param start.c optional initial guess for the third curve fit coefficient. Either \code{c} in case of \code{Inverse_quad} or \code{xsd} in case of \code{Bell}. Not used for \code{Wait} as it only uses two coefficients. Default is NA, i.e. an internal guess is used to determine \code{start.c}.
#' @seealso \code{nls}
#' @examples
#' d <- data.frame(x = 1:10,
#'                 y = c(0.5, 2, 10, 9, 8, 6.5, 5, 4, 3, 2.5))
#' MyFits(x = x, y = y, data = d)
#' MyFits(x = x, y = y, data = d, start.a = 1.5, start.b = 0.5, start.c = 0.05)
#' MyFits(x = x, y = y, data = d, fit = "Wait")
#' nls.fit <- MyFits(x = x, y = y, data = d, fit = "Inverse_quad", model = TRUE)
#' my.x <- seq(min(d$x), max(d$x), length.out = 100)
#' my.predicted <- predict(nls.fit, data.frame(x = my.x))
#' mod.pred <- data.frame(x = my.x,
#'                        y = my.predicted)
#' plot(d)
#' lines(y ~ x, data = mod.pred)
#' @export


MyFits <- function(x, y, data, 
                   fit = "Inverse_quad",
                   model = FALSE, coords = FALSE,
                   start.a = NA, start.b = NA, start.c = NA) {

  stopifnot(fit %in% c("Inverse_quad", "Bell", "Wait"))
  
  arguments <- as.list(match.call())
  x <- eval(arguments$x, data)
  y <- eval(arguments$y, data)
  
  # make sure y, y are numeric, as it could be time 
  x <- as.numeric(x)
  y <- as.numeric(y)
  
  
  if (is.na(start.a) & is.na(start.b) & is.na(start.b)) {
    message("Determining start values")
    
    if (fit == "Inverse_quad") {
      message(fit)
      my.eq   <- stats::formula(y ~ x / (a + b*x + c*x^2))
      start.a <- min(x[x>0], na.rm = TRUE) / 10
      start.b <- max(y,      na.rm = TRUE) 
      start.c <- 1/start.a
      start.para <- c(start.a, start.b, start.c)
      names(start.para) <- c("a", "b", "c")
    }
    
    if (fit == "Bell") {
      message(fit)
      my.eq <- stats::formula(y ~ amp * exp(-0.5*((x - xmean)/xsd)^2))
      start.amp   <- max(y,  na.rm = TRUE) 
      start.xmean <- mean(x, na.rm = TRUE)
      start.xsd   <- stats::sd(x,   na.rm = TRUE)
      start.para  <- c(start.amp, start.xmean, start.xsd)
      names(start.para) <- c("amp", "xmean", "xsd")
    }
    
    if (fit == "Wait") {
      message(fit)
      my.eq    <- stats::formula(y ~ A * kw * x * exp(-kw*x))
      start.A  <- max(y, na.rm = TRUE) # scaling constant for height of curve
      # find x-value of maximum
      x.max      <- x[which.max(y)]
      start.kw   <- 1/x.max # location of saturation
      start.para <- c(start.A, start.kw)
      names(start.para) <- c("A", "kw")
    }
    
  } else {
    message("Using provided start values")
    if (fit == "Inverse_quad") {
      message(fit)
      my.eq   <- stats::formula(y ~ x / (a + b*x + c*x^2))
      start.a <- start.a
      start.b <- start.b
      start.c <- start.b
      start.para <- c(start.a, start.b, start.c)
      names(start.para) <- c("a", "b", "c")
    }
    
    if (fit == "Bell") {
      message(fit)
      my.eq <- stats::formula(y ~ amp * exp(-0.5*((x - xmean)/xsd)^2))
      start.amp   <- start.a
      start.xmean <- start.b
      start.xsd   <- start.c
      start.para  <- c(start.amp, start.xmean, start.xsd)
      names(start.para) <- c("amp", "xmean", "xsd")
    }
    
    if (fit == "Wait") {
      message(fit)
      my.eq    <- stats::formula(y ~ A * kw * x * exp(-kw*x))
      start.A  <- start.a
      start.kw <- start.b
      start.para <- c(start.A, start.kw)
      names(start.para) <- c("A", "kw")
    }
  }
    
  the.names <- c(names(start.para), "R2")
  
  nls.fit <- try(stats::nls(my.eq,
                     #data = data,
                     trace = FALSE, 
                     start = start.para,
                     control = stats::nls.control(minFactor = 1/40000, 
                                           maxiter = 8000)))

  if(inherits(nls.fit, "try-error")) {
     out <- rep(NA, length(the.names))
     names(out) <- the.names
     if (model == TRUE) {
       message("No fit")
       return("No fit")
     } else {
       return(out)
     }
  } else {
   if (model == TRUE) {
     if (coords == FALSE) {
       # return the bare model
       return(nls.fit)
     } else {
      # do prediction
      my.x <- seq(min(x, na.rm = TRUE), 
                  max(x, na.rm = TRUE), 
                  length.out = length(x) * 4)
      my.predicted <- stats::predict(nls.fit, data.frame(x = my.x))
      mod.pred <- data.frame(x = my.x,
                             y = my.predicted)
      return(mod.pred)
     }
   } else {
     # provide coefficients and R2
     out <- stats::coef(nls.fit)
     r2  <- 1 - (stats::var(stats::residuals(nls.fit)) / stats::var(y, na.rm = TRUE))
     out <- c(out, r2)
     names(out) <- the.names
     return(out)
     }
  }
}



R/stat_MyFits.R
#' Use MyFits to add a curve fits to a ggplot2 object.
#' 
#' @description Calculates either an inverse quadratic, a bell-shaped or a waiting-in-line fit of data as a \code{ggplot2} layer. Calculated using \code{MyFits::MyFits}.
#' @param fit Name of the fit-equation to be used. Either \code{Inverse_quad} for an inverse quadratic equation \code{y ~ x / (a + b*x + c*x^2)}, \code{Bell} for a bell shaped curve \code{y ~ amp * exp(-0.5*((x - xmean)/xsd)^2)}, or \code{Wait} for waiting-in-line \code{y ~ A * kw * x * exp(-kw*x)}. See \code{MyFits}.
#' @param start.a optional initial guess for the curve fit coefficients \code{a}, \code{amp}, or \code{A} (depends on which equation was selected). Default is NA, i.e. an internal guess is used to determine \code{start.a}.
#' @param start.b optional initial guess for the second curve fit coefficient for the \code{b}, \code{xmean} or \code{kw} parameter. Default is NA, i.e. an internal guess is used to determine \code{start.b}.
#' @param start.c optional initial guess for the third curve fit coefficient. Either \code{c} in case of \code{Inverse_quad} or \code{xsd} in case of \code{Bell}. Not used for \code{Wait} as it only uses two coefficients. Default is NA, i.e. an internal guess is used to determine \code{start.c}.
# '@inheritParams StatMyFits
#' @inheritParams ggplot2::stat_summary
#' @inheritParams ggplot2::layer
#' @seealso  \code{MyFits}, \code{ggplot2}
#' @examples
#' d <- data.frame(x = 1:10,
#'                 y = c(0.5, 2, 10, 9, 8, 6.5, 5, 4, 3, 2.5))
#' 
#' require(ggplot2)
#' p <- ggplot(d, aes(x = x, y = y))
#'  p <- p + geom_point()
#'  p <- p + stat_myfits(fit = "Inverse_quad", colour = "blue")
#'  p <- p + stat_myfits(fit = "Bell", colour = "red")
#'  p <- p + stat_myfits(fit = "Wait", colour = "green")
#' p 
#' 
#' # providing inital guesses for the start values.
#' p <- ggplot(d, aes(x = x, y = y))
#'   p <- p + geom_point()
#'   p <- p + stat_myfits(fit = "Bell", colour = "blue", 
#'            start.a = 10, start.b = 1, start.c = 5)
#' p
#' 
#' @export stat_myfits

stat_myfits <- function(mapping = NULL, data = NULL, geom = "path",
                        position = "identity", na.rm = FALSE, show.legend = NA, 
                        inherit.aes = TRUE, fit, 
                        start.a = NA, start.b = NA, start.c = NA, ...) {
  ggplot2::layer(
    stat = StatMyFits, data = data, mapping = mapping, geom = geom, 
    position = position, show.legend = show.legend, inherit.aes = inherit.aes, 
    params = list(na.rm = na.rm, fit = fit, 
    start.a = start.a, start.b = start.a, start.c = start.c,...)
  )
}

StatMyFits <- ggplot2::ggproto("StatMyFits", ggplot2::Stat, 
  required_aes = c("x", "y"),
  
  compute_group = function(data, scales, fit = fit, 
                           start.a = start.a, 
                           start.b = start.b, 
                           start.c = start.c) {
    MyFits(x = data$x, y = data$y, data = data, fit, 
           model = TRUE, coords = TRUE, 
           start.a = start.a, start.b = start.b, start.c = start.c)
  }
)

